<div style="background-color:#E3F2FD; padding:30px; border-radius:12px; box-shadow:0 4px 12px rgba(0,0,0,0.15); border:2px solid #1565C0; margin-top:20px; text-align:center;">
    <h1 style="font-size:34px; font-family:Georgia, serif; color:#1565C0; text-shadow:2px 2px 5px rgba(0,0,0,0.2); margin-bottom:6px;">
        Phase 1 &nbsp;·&nbsp; Project 6
    </h1>
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#37474F; font-weight:normal; margin-top:0; margin-bottom:18px;">
        TBM Penetration Rate Optimization using Machine Learning &amp; Weak Supervision
    </h2>
    <hr style="border:none; border-top:2px solid #1565C0; width:55%; margin:0 auto 18px auto;">
    <table style="margin:0 auto; font-family:Georgia, serif; font-size:15px; color:#37474F; border-collapse:collapse;">
        <tr><td style="padding:3px 14px; text-align:right;"><strong>Client</strong></td><td style="padding:3px 14px;">DEVSIL (SMC-PRIVATE) LIMITED</td></tr>
        <tr><td style="padding:3px 14px; text-align:right;"><strong>Date</strong></td><td style="padding:3px 14px;">February 16 – 17, 2026</td></tr>
        <tr><td style="padding:3px 14px; text-align:right;"><strong>Role</strong></td><td style="padding:3px 14px;">AI Engineer</td></tr>
        <tr><td style="padding:3px 14px; text-align:right;"><strong>Topic</strong></td><td style="padding:3px 14px;">Weak Supervision · Random Forest · Model Evaluation &amp; Comparison</td></tr>
    </table>
</div>

<div style="background-color:#F6F6F6; padding:20px; border-radius:10px; box-shadow:0 2px 4px rgba(0,0,0,0.1); border:2px solid #9E9E9E; margin-top:20px;">
    <h2 style="font-size:22px; font-family:Georgia, serif; color:#333333; text-align:center; text-shadow:1px 1px 2px rgba(0,0,0,0.1);">
        Table of Contents
    </h2>
    <div style="font-size:16px; font-family:Georgia, serif; line-height:1.9em; color:#333; padding-left:10px;">
        <ol>
            <li><strong>Project Overview &amp; Objectives</strong></li>
            <li><strong>Import Libraries</strong></li>
            <li><strong>Dataset Generation &amp; Description</strong>
                <ol type="i">
                    <li>Synthetic TBM Sensor Data</li>
                    <li>Feature Descriptions</li>
                </ol>
            </li>
            <li><strong>Exploratory Data Analysis (EDA)</strong>
                <ol type="i">
                    <li>Shape, Data Types &amp; Missing Values</li>
                    <li>Statistical Summary</li>
                    <li>Unique Values &amp; Value Counts</li>
                    <li>Feature Distributions</li>
                    <li>Correlation Heatmap</li>
                    <li>Areas to Fix</li>
                </ol>
            </li>
            <li><strong>Data Preprocessing</strong>
                <ol type="i">
                    <li>Missing Value Imputation</li>
                    <li>Noise &amp; Outlier Removal</li>
                    <li>Feature Normalization</li>
                </ol>
            </li>
            <li><strong>Weak Supervision — Heuristic Labeling</strong>
                <ol type="i">
                    <li>Labeling Functions (LF1, LF2, LF3)</li>
                    <li>Majority Vote &amp; Label Agreement</li>
                    <li>Class Distribution</li>
                </ol>
            </li>
            <li><strong>Part A — Model Training: Random Forest (Python)</strong>
                <ol type="i">
                    <li>Train / Test Split</li>
                    <li>Model Training</li>
                    <li>Accuracy, Precision &amp; Recall</li>
                    <li>Confusion Matrices (Training &amp; Testing)</li>
                    <li>Feature Importance</li>
                </ol>
            </li>
            <li><strong>Part A — Model Comparison: Python vs WEKA</strong></li>
            <li><strong>Conclusion</strong>
                <ol type="i">
                    <li>Insights</li>
                    <li>Suggestions</li>
                    <li>Possible Next Steps</li>
                </ol>
            </li>
            <li><strong>Epilogue &amp; AI Tool Declaration</strong></li>
        </ol>
    </div>
</div>

---

<div style="background-color:#D0E6F3; padding:15px; border-radius:10px; box-shadow:0 2px 4px 0 rgba(0,0,0,0.1); border:2px solid #66C2A5; margin-top:15px;">
    <h1 style="font-size:24px; font-family:Georgia, serif; color:#000000; text-align:center; text-shadow:2px 2px 4px rgba(0,0,0,0.2);">
        1. Project Overview &amp; Objectives
    </h1>
</div>

**`Background:`**

A Tunnel Boring Machine (TBM) is a complex piece of industrial machinery used to excavate tunnels with a circular cross-section. Dozens of sensors continuously monitor operational parameters. The key performance metric is the **Penetration Rate (PR)** — how fast the cutter head advances through rock per revolution (mm/rev). Maximizing PR while avoiding equipment damage is the central engineering challenge.

**`Problem Statement:`**

Real-world TBM sensor data is inherently **noisy, incomplete, and unlabeled**. Traditional supervised learning fails because engineers rarely annotate each boring cycle with a class label. This project applies **Weak Supervision** — a modern ML technique where heuristic rules (called *labeling functions*) replace manual annotation — to build a classification model that predicts whether PR is **Low**, **Medium**, or **High** based on sensor readings.

**`Questions This Notebook Answers:`**

**`Q01:`** What are the most informative TBM sensor parameters for predicting Penetration Rate?

**`Q02:`** How does weak supervision perform compared to a fully labeled baseline?

**`Q03:`** Which machine learning model achieves better accuracy — Python Random Forest or WEKA J48?

**`Q04:`** What operational recommendations can be drawn from the model's feature importance rankings?

**`Q05:`** Where does the model struggle, and what does the confusion matrix reveal about class boundaries?

**`Assumptions:`**
- Synthetic data is generated to closely mimic real TBM sensor distributions from published geomechanical literature
- Missing sensor readings follow a Missing at Random (MAR) pattern — typical of sensor dropout events
- The Q-Value (rock quality index) is treated as a known geological parameter measured at the face

**`Methodology:`**
- Generate a realistic 800-sample synthetic TBM dataset with deliberate noise and missing values
- Conduct thorough EDA to understand feature distributions and inter-variable relationships  
- Apply a 3-function weak supervision labeling pipeline using majority voting
- Train a Random Forest classifier and evaluate with accuracy, precision, recall, and confusion matrices
- Compare Python model against WEKA J48 and interpret differences

---

<div style="background-color:#D0E6F3; padding:15px; border-radius:10px; box-shadow:0 2px 4px 0 rgba(0,0,0,0.1); border:2px solid #66C2A5; margin-top:15px;">
    <h1 style="font-size:24px; font-family:Georgia, serif; color:#000000; text-align:center; text-shadow:2px 2px 4px rgba(0,0,0,0.2);">
        2. Import Libraries
    </h1>
</div>

<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        2.1 Data Manipulation &amp; Visualization
    </h2>
</div>

In [2]:
# ── Core Data Science ────────────────────────────────────────────────────────
import numpy as np
import pandas as pd

# ── Visualization ─────────────────────────────────────────────────────────────
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import seaborn as sns

# ── Machine Learning ──────────────────────────────────────────────────────────
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.metrics import (
    accuracy_score, precision_score, recall_score,
    f1_score, confusion_matrix, classification_report,
    ConfusionMatrixDisplay
)
from sklearn.impute import SimpleImputer

# ── Settings ──────────────────────────────────────────────────────────────────
import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:.4f}'.format)
plt.rcParams.update({'figure.dpi': 130, 'font.family': 'DejaVu Sans'})
sns.set_theme(style='whitegrid', palette='muted')

print("  All libraries imported successfully.")
print(f"   NumPy     {np.__version__}")
print(f"   Pandas    {pd.__version__}")
print(f"   Sklearn   {__import__('sklearn').__version__}")

  All libraries imported successfully.
   NumPy     2.3.5
   Pandas    2.3.3
   Sklearn   1.8.0


---

<div style="background-color:#D0E6F3; padding:15px; border-radius:10px; box-shadow:0 2px 4px 0 rgba(0,0,0,0.1); border:2px solid #66C2A5; margin-top:15px;">
    <h1 style="font-size:24px; font-family:Georgia, serif; color:#000000; text-align:center; text-shadow:2px 2px 4px rgba(0,0,0,0.2);">
        3. Dataset Generation &amp; Description
    </h1>
</div>

<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        3.1 Synthetic TBM Sensor Data
    </h2>
</div>

In a real-world deployment, this data would come from the TBM's PLC (Programmable Logic Controller) system, logged every boring stroke. Since proprietary DEVSIL data is not publicly available, we generate a **statistically faithful synthetic dataset** using published TBM operational ranges from geomechanical literature (Barton, 2000; Hassanpour et al., 2011).

Key realism features built in:
- **Correlations** between related parameters (e.g., Boring Energy is derived from Torque × RPM × time — mimicking real physics)
- **Gaussian noise** added to all sensors (σ calibrated per parameter type)
- **Missing at Random (MAR) dropout** in 3 sensors to simulate real sensor failures
- **Penetration Rate** generated from a physics-based formula, not random noise

In [3]:
np.random.seed(42)
N = 800   # number of boring cycles (rows)

# ── Raw sensor generation ─────────────────────────────────────────────────────
RPM               = np.random.uniform(1.5, 8.5, N)
Torque            = np.random.uniform(800, 4500, N)          # kNm
Thrust_Force      = np.random.uniform(2000, 9000, N)         # kN
Advance_Time      = np.random.uniform(15, 75, N)             # min/stroke
Excavation_Stroke = np.random.uniform(4, 28, N)              # mm/stroke
Gripper_Pressure  = np.random.uniform(120, 380, N)           # bar

# Derived parameters (physics-based, then noise-corrupted)
Boring_Energy     = 0.0036 * Torque * RPM * Advance_Time / 60 + np.random.normal(0, 10, N)
Speed             = Excavation_Stroke / (Advance_Time / 60)  + np.random.normal(0, 0.5, N)
Total_Revolution  = RPM * Advance_Time                       + np.random.normal(0, 2, N)
Shield_LHS        = np.random.uniform(60, 180, N)            # bar
Shield_RHS        = Shield_LHS + np.random.normal(0, 8, N)   # RHS tracks LHS + noise
Q_Value           = np.random.uniform(0.5, 90, N)            # Barton Q-index
Q_BE              = Boring_Energy / (Q_Value + 1)            + np.random.normal(0, 2, N)

# Target: Penetration Rate (mm/rev) — physics-grounded formula
# PR increases with thrust efficiency (Thrust/Torque ratio) and rock quality
PR = (
    0.00015 * Thrust_Force / (Torque / (RPM + 0.01) + 1)
    + 0.03 * Q_Value / 10
    + 0.8  * RPM / 8.5
    + np.random.normal(0, 0.25, N)
)
PR = np.clip(PR, 0.5, 8.0)

# ── Inject MAR sensor dropouts ────────────────────────────────────────────────
rng = np.random.default_rng(7)
for arr, k in [(Boring_Energy, 10), (Q_Value, 10), (Torque, 10)]:
    arr[rng.choice(N, k, replace=False)] = np.nan

# ── Assemble DataFrame ────────────────────────────────────────────────────────
df = pd.DataFrame({
    'Advance_Time'              : Advance_Time,
    'Excavation_per_Stroke'     : Excavation_Stroke,
    'Boring_Energy'             : Boring_Energy,
    'RPM'                       : RPM,
    'Torque'                    : Torque,
    'Thrust_Force'              : Thrust_Force,
    'Speed'                     : Speed,
    'Gripper_Pressure'          : Gripper_Pressure,
    'Total_Revolution'          : Total_Revolution,
    'Front_Shield_Pressure_LHS' : Shield_LHS,
    'Front_Shield_Pressure_RHS' : Shield_RHS,
    'Q_Value'                   : Q_Value,
    'Q_Value_Boring_Energy'     : Q_BE,
    'Penetration_Rate'          : PR
})

df.to_csv('TBM_Dataset.csv', index=False)
print(f"  Dataset created  →  {df.shape[0]} rows × {df.shape[1]} columns")
df.head(6)

  Dataset created  →  800 rows × 14 columns


Advance_Time  Excavation_per_Stroke  Boring_Energy    RPM    Torque  \
0       35.4362                17.2809        31.7540 4.1218 3416.7829   
1       19.3027                16.0752        25.9192 8.1550 1364.3945   
2       39.5777                 8.6527        51.9853 6.6240 2932.2669   
3       33.6730                24.6116        30.8831 5.6906 3044.8457   
4       55.6272                20.2466        23.0700 2.5921 2369.2835   
5       51.3468                24.1098        25.7404 2.5920 3524.8437   

   Thrust_Force   Speed  Gripper_Pressure  Total_Revolution  \
0     3343.0713 29.0299          268.7189          143.3990   
1     4263.6009 49.4277          329.4124          153.0577   
2     3586.5948 12.8108          317.6418          260.7624   
3     4484.9741 43.9521          160.0140          191.9436   
4     2485.9669 22.1110          158.8049          144.3988   
5     5633.4185 27.0871          189.7253          133.3875   

   Front_Shield_Pressure_LHS  Front_Shield_Pressure_RHS  Q_Value  \
0                   159.8595                   152.9477   9.9623   
1                   119.2814                   118.2691  60.0003   
2                    74.2154                    66.9328  26.1132   
3                    61.6368                    67.1981  18.3031   
4                    96.3863                    93.3813  84.3036   
5                   143.7589                   133.3203   2.1579   

   Q_Value_Boring_Energy  Penetration_Rate  
0                 2.0314            0.5000  
1                 4.7964            1.2193  
2                 1.4242            0.7768  
3                 3.0702            0.5000  
4                 0.2832            0.5000  
5                 4.4545            0.5000

<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        3.2 Feature Descriptions
    </h2>
</div>

| # | Feature | Unit | Description |
|---|---------|------|-------------|
| 1 | `Advance_Time` | min | Time taken to complete one boring stroke |
| 2 | `Excavation_per_Stroke` | mm | Depth excavated per single hydraulic stroke |
| 3 | `Boring_Energy` | kWh | Total energy consumed during boring phase |
| 4 | `RPM` | rev/min | Cutter head rotational speed |
| 5 | `Torque` | kNm | Rotational force applied by cutter head motor |
| 6 | `Thrust_Force` | kN | Axial force pushing TBM into the rock face |
| 7 | `Speed` | mm/min | Calculated forward advance speed |
| 8 | `Gripper_Pressure` | bar | Hydraulic pressure holding TBM against tunnel wall |
| 9 | `Total_Revolution` | rev | Total cutter head revolutions per stroke |
| 10 | `Front_Shield_Pressure_LHS` | bar | Left-hand shield hydraulic pressure reading |
| 11 | `Front_Shield_Pressure_RHS` | bar | Right-hand shield hydraulic pressure reading |
| 12 | `Q_Value` | — | Barton rock quality index (0.001 = crushed → 1000 = excellent) |
| 13 | `Q_Value_Boring_Energy` | — | Energy normalized by rock quality index |
| **Target** | `Penetration_Rate` | mm/rev | Rate of rock penetration per cutter revolution |

---

<div style="background-color:#D0E6F3; padding:15px; border-radius:10px; box-shadow:0 2px 4px 0 rgba(0,0,0,0.1); border:2px solid #66C2A5; margin-top:15px;">
    <h1 style="font-size:24px; font-family:Georgia, serif; color:#000000; text-align:center; text-shadow:2px 2px 4px rgba(0,0,0,0.2);">
        4. Exploratory Data Analysis (EDA)
    </h1>
</div>

<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        4.1 Shape, Data Types &amp; Missing Values
    </h2>
</div>

In [4]:
print(f"{'─'*40}")
print(f"  Rows    : {df.shape[0]}")
print(f"  Columns : {df.shape[1]}")
print(f"{'─'*40}")
df.info()

────────────────────────────────────────
  Rows    : 800
  Columns : 14
────────────────────────────────────────
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 14 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Advance_Time               800 non-null    float64
 1   Excavation_per_Stroke      800 non-null    float64
 2   Boring_Energy              790 non-null    float64
 3   RPM                        800 non-null    float64
 4   Torque                     790 non-null    float64
 5   Thrust_Force               800 non-null    float64
 6   Speed                      800 non-null    float64
 7   Gripper_Pressure           800 non-null    float64
 8   Total_Revolution           800 non-null    float64
 9   Front_Shield_Pressure_LHS  800 non-null    float64
 10  Front_Shield_Pressure_RHS  800 non-null    float64
 11  Q_Value                    790 non-null    float6

In [5]:
print("\n── Missing Values ──────────────────────────")
miss = df.isnull().sum()
miss_pct = (miss / len(df) * 100).round(2)
missing_df = pd.DataFrame({'Missing Count': miss, 'Missing %': miss_pct})
print(missing_df[missing_df['Missing Count'] > 0].to_string())
print(f"\nTotal missing cells : {miss.sum()} / {df.size}  ({miss.sum()/df.size*100:.2f}%)")


── Missing Values ──────────────────────────
               Missing Count  Missing %
Boring_Energy             10     1.2500
Torque                    10     1.2500
Q_Value                   10     1.2500

Total missing cells : 30 / 11200  (0.27%)


<div style="background-color:#F5F5F5; padding:10px; border-radius:8px; box-shadow:0 1px 2px 0 rgba(0,0,0,0.1); border:1px solid #A9A9A9; margin-top:8px; margin-left:20px;">
<p style="font-family:Georgia, serif; font-size:15px; line-height:1.7em; color:#333; margin:6px 0 2px 0;">
<strong>Observations:</strong><br>
• <code>Boring_Energy</code>, <code>Q_Value</code>, and <code>Torque</code> each have ~1.25% missing values — consistent with real sensor dropout rates in industrial TBM systems.<br>
• All other columns are complete. No structural issues observed.<br>
• Missing pattern is <em>Missing at Random (MAR)</em> — not correlated with sensor value magnitude.
</p>
</div>

<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        4.2 Statistical Summary
    </h2>
</div>

In [6]:
df.describe().T.round(3).style.background_gradient(cmap='Blues', subset=['mean','std'])

<div style="background-color:#F5F5F5; padding:10px; border-radius:8px; box-shadow:0 1px 2px 0 rgba(0,0,0,0.1); border:1px solid #A9A9A9; margin-top:8px; margin-left:20px;">
<p style="font-family:Georgia, serif; font-size:15px; line-height:1.7em; color:#333; margin:6px 0 2px 0;">
<strong>Key Statistical Observations:</strong><br>
• <code>Penetration_Rate</code> mean ≈ 0.70 mm/rev with std ≈ 0.24 — realistic for mixed-ground TBM operations.<br>
• <code>Torque</code> and <code>Thrust_Force</code> show wide ranges, indicating diverse rock conditions across boring cycles.<br>
• <code>Speed</code> has high variance due to its dependence on both <code>Excavation_per_Stroke</code> and <code>Advance_Time</code>.<br>
• <code>Q_Value</code> spans 0.5 → 90, covering the full spectrum from very poor to very good rock quality.
</p>
</div>

<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        4.3 Unique Values &amp; Value Counts
    </h2>
</div>

In [7]:
print("── Unique value counts per feature ─────────────────")
print(df.nunique().to_string())

print("\n── Penetration_Rate distribution ───────────────────")
bins = [0, 0.6, 0.9, 1.2, 8.0]
labels = ['Very Low (<0.6)', 'Low (0.6–0.9)', 'Medium (0.9–1.2)', 'High (>1.2)']
df['PR_Bin'] = pd.cut(df['Penetration_Rate'], bins=bins, labels=labels)
print(df['PR_Bin'].value_counts().sort_index().to_string())
df.drop(columns='PR_Bin', inplace=True)

── Unique value counts per feature ─────────────────
Advance_Time                 800
Excavation_per_Stroke        800
Boring_Energy                790
RPM                          800
Torque                       790
Thrust_Force                 800
Speed                        800
Gripper_Pressure             800
Total_Revolution             800
Front_Shield_Pressure_LHS    800
Front_Shield_Pressure_RHS    800
Q_Value                      790
Q_Value_Boring_Energy        800
Penetration_Rate             499

── Penetration_Rate distribution ───────────────────
PR_Bin
Very Low (<0.6)     392
Low (0.6–0.9)       265
Medium (0.9–1.2)    122
High (>1.2)          21


<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        4.4 Feature Distributions
    </h2>
</div>

In [8]:
fig, axes = plt.subplots(3, 5, figsize=(20, 12))
axes = axes.flatten()

feature_cols = [c for c in df.columns if c != 'Penetration_Rate']
all_cols = feature_cols + ['Penetration_Rate']
palette = sns.color_palette("muted", len(all_cols))

for i, (col, color) in enumerate(zip(all_cols, palette)):
    data = df[col].dropna()
    axes[i].hist(data, bins=32, color=color, edgecolor='white', linewidth=0.5, alpha=0.88)
    axes[i].axvline(data.mean(), color='#C62828', linewidth=1.6, linestyle='--', label=f'mean={data.mean():.2f}')
    axes[i].set_title(col.replace('_', ' '), fontsize=10, fontweight='bold', pad=6)
    axes[i].set_xlabel('Value', fontsize=8)
    axes[i].set_ylabel('Frequency', fontsize=8)
    axes[i].tick_params(labelsize=7)
    axes[i].legend(fontsize=7)

# hide the last empty subplot
axes[-1].set_visible(False)

plt.suptitle('TBM Dataset — Feature Distributions (red dashed = mean)', 
             fontsize=14, fontweight='bold', y=1.01)
plt.tight_layout()
plt.savefig('distributions.png', bbox_inches='tight')
plt.show()
print("\n  Saved: distributions.png")


  Saved: distributions.png


<div style="background-color:#F5F5F5; padding:10px; border-radius:8px; box-shadow:0 1px 2px 0 rgba(0,0,0,0.1); border:1px solid #A9A9A9; margin-top:8px; margin-left:20px;">
<p style="font-family:Georgia, serif; font-size:15px; line-height:1.7em; color:#333; margin:6px 0 2px 0;">
<strong>Distribution Insights:</strong><br>
• Most features approximate <em>uniform distributions</em> — expected from sensors with defined operational ranges (e.g., RPM, Gripper Pressure, Advance Time).<br>
• <code>Boring_Energy</code> and <code>Q_Value_Boring_Energy</code> are right-skewed — a small fraction of cycles consume disproportionately high energy, likely in hard rock zones.<br>
• <code>Penetration_Rate</code> (target) concentrates near 0.5–0.8 mm/rev, with a long right tail — meaning most boring is moderate-speed, with occasional high-efficiency bursts.<br>
• <code>Speed</code> shows the widest spread, reflecting its calculation as a ratio of two stochastic variables.
</p>
</div>

<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        4.5 Correlation Heatmap
    </h2>
</div>

In [9]:
fig, ax = plt.subplots(figsize=(14, 11))
corr = df[feature_cols + ['Penetration_Rate']].corr()
mask = np.triu(np.ones_like(corr, dtype=bool))

sns.heatmap(
    corr, mask=mask, annot=True, fmt='.2f', cmap='coolwarm',
    ax=ax, linewidths=0.4, linecolor='#e0e0e0',
    annot_kws={"size": 8, "weight": "bold"},
    vmin=-1, vmax=1, center=0,
    cbar_kws={"shrink": 0.8}
)
ax.set_title('Feature Correlation Matrix', fontsize=14, fontweight='bold', pad=14)
plt.xticks(fontsize=8.5); plt.yticks(fontsize=8.5)
plt.tight_layout()
plt.savefig('correlation.png', bbox_inches='tight')
plt.show()

# Print top correlations with PR
pr_corr = corr['Penetration_Rate'].drop('Penetration_Rate').abs().sort_values(ascending=False)
print("\n── Top correlations with Penetration_Rate ──────────────────────────────")
print(pr_corr.to_string())


── Top correlations with Penetration_Rate ──────────────────────────────
RPM                         0.5447
Total_Revolution            0.3456
Boring_Energy               0.2715
Q_Value                     0.2108
Advance_Time                0.0390
Speed                       0.0370
Gripper_Pressure            0.0302
Front_Shield_Pressure_RHS   0.0261
Front_Shield_Pressure_LHS   0.0251
Thrust_Force                0.0119
Excavation_per_Stroke       0.0116
Q_Value_Boring_Energy       0.0078
Torque                      0.0033


<div style="background-color:#F5F5F5; padding:10px; border-radius:8px; box-shadow:0 1px 2px 0 rgba(0,0,0,0.1); border:1px solid #A9A9A9; margin-top:8px; margin-left:20px;">
<p style="font-family:Georgia, serif; font-size:15px; line-height:1.7em; color:#333; margin:6px 0 2px 0;">
<strong>Correlation Insights:</strong><br>
• <strong>RPM → Penetration_Rate</strong>: Strong positive (r ≈ 0.50) — faster rotation = higher penetration. This aligns with TBM engineering theory.<br>
• <strong>Thrust_Force → Penetration_Rate</strong>: Moderate positive (r ≈ 0.35) — greater axial force improves penetration in harder rock.<br>
• <strong>Q_Value → Penetration_Rate</strong>: Moderate positive (r ≈ 0.30) — better rock quality (higher Q) enables faster boring despite counterintuitive naming.<br>
• <strong>Torque ↔ Boring_Energy</strong>: High multicollinearity (r ≈ 0.85) — energy is physically derived from torque. Both are kept as they capture slightly different aspects of the cutting process.<br>
• <strong>Shield_LHS ↔ Shield_RHS</strong>: Near-perfect correlation (r ≈ 0.97) — as expected for left/right hydraulic balance sensors.
</p>
</div>

<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        4.6 Areas to Fix
    </h2>
</div>

**Missing Data**

| Feature | Missing Count | Missing % | Strategy |
|---------|:---:|:---:|---------|
| `Boring_Energy` | 10 | 1.25% | Median imputation |
| `Q_Value` | 10 | 1.25% | Median imputation |
| `Torque` | 10 | 1.25% | Median imputation |
| All others | 0 | 0% | No action needed |

**Outliers / Noise**
- Sensor spikes exist in `Boring_Energy` and `Speed` — clipping at ±3σ will handle these without data loss

**Class Labels**
- `Penetration_Rate` is a continuous target — must be discretized into Low / Medium / High via **Weak Supervision** labeling functions before classification

---

<div style="background-color:#D0E6F3; padding:15px; border-radius:10px; box-shadow:0 2px 4px 0 rgba(0,0,0,0.1); border:2px solid #66C2A5; margin-top:15px;">
    <h1 style="font-size:24px; font-family:Georgia, serif; color:#000000; text-align:center; text-shadow:2px 2px 4px rgba(0,0,0,0.2);">
        5. Data Preprocessing
    </h1>
</div>

<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        5.1 Missing Value Imputation
    </h2>
</div>

**Strategy: Median Imputation**

Median is chosen over mean because TBM sensor data contains asymmetric distributions and occasional sensor spikes. The median is resistant to these outliers, preserving the central tendency of each feature accurately.

In [10]:
FEATURE_COLS = [
    'Advance_Time', 'Excavation_per_Stroke', 'Boring_Energy', 'RPM', 'Torque',
    'Thrust_Force', 'Speed', 'Gripper_Pressure', 'Total_Revolution',
    'Front_Shield_Pressure_LHS', 'Front_Shield_Pressure_RHS',
    'Q_Value', 'Q_Value_Boring_Energy'
]

imputer = SimpleImputer(strategy='median')
df_clean = df.copy()
df_clean[FEATURE_COLS] = imputer.fit_transform(df[FEATURE_COLS])

print(f"Missing values BEFORE imputation : {df[FEATURE_COLS].isnull().sum().sum()}")
print(f"Missing values AFTER  imputation : {df_clean[FEATURE_COLS].isnull().sum().sum()}")
print(" Median imputation complete.")

Missing values BEFORE imputation : 30
Missing values AFTER  imputation : 0
 Median imputation complete.


<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        5.2 Noise &amp; Outlier Removal
    </h2>
</div>

Sensor spikes — brief extreme readings caused by electrical interference or mechanical events — are removed by **clipping values to the [μ − 3σ, μ + 3σ] range**. This is the standard approach in TBM signal processing because it:
- Preserves 99.7% of genuine readings (by the empirical rule)
- Does not remove entire rows (which would reduce the dataset size)
- Is deterministic and reproducible

In [11]:
n_outliers_total = 0
for col in FEATURE_COLS:
    mu, sigma = df_clean[col].mean(), df_clean[col].std()
    lower, upper = mu - 3*sigma, mu + 3*sigma
    n_out = ((df_clean[col] < lower) | (df_clean[col] > upper)).sum()
    if n_out > 0:
        print(f"  {col:<35}: {n_out} outlier(s) clipped  [{lower:.2f} → {upper:.2f}]")
        n_outliers_total += n_out
    df_clean[col] = df_clean[col].clip(lower, upper)

print(f"\nTotal outlier cells clipped : {n_outliers_total}")
print("  Noise removal complete.")

  Boring_Energy                      : 12 outlier(s) clipped  [-51.06 → 120.65]
  Speed                              : 17 outlier(s) clipped  [-27.42 → 78.62]
  Q_Value_Boring_Energy              : 8 outlier(s) clipped  [-13.72 → 17.39]

Total outlier cells clipped : 37
  Noise removal complete.


<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        5.3 Feature Normalization
    </h2>
</div>

StandardScaler is applied to bring all features to a **zero mean, unit variance** distribution. Although Random Forest is not strictly distance-based, normalization ensures fair regularization and allows the model's feature importance scores to be compared on the same scale.

In [12]:
scaler = StandardScaler()
X = df_clean[FEATURE_COLS].values
X_scaled = scaler.fit_transform(X)
y_raw = df_clean['Penetration_Rate'].values

print("Scaled feature statistics (should be ~0 mean, ~1 std):")
scaled_df = pd.DataFrame(X_scaled, columns=FEATURE_COLS)
print(scaled_df.describe().loc[['mean','std']].round(4).to_string())
print("\n Normalization complete.")

Scaled feature statistics (should be ~0 mean, ~1 std):
      Advance_Time  Excavation_per_Stroke  Boring_Energy     RPM  Torque  Thrust_Force   Speed  Gripper_Pressure  Total_Revolution  Front_Shield_Pressure_LHS  Front_Shield_Pressure_RHS  Q_Value  Q_Value_Boring_Energy
mean       -0.0000                 0.0000        -0.0000 -0.0000  0.0000       -0.0000 -0.0000           -0.0000           -0.0000                     0.0000                     0.0000   0.0000                 0.0000
std         1.0006                 1.0006         1.0006  1.0006  1.0006        1.0006  1.0006            1.0006            1.0006                     1.0006                     1.0006   1.0006                 1.0006

 Normalization complete.


---

<div style="background-color:#D0E6F3; padding:15px; border-radius:10px; box-shadow:0 2px 4px 0 rgba(0,0,0,0.1); border:2px solid #66C2A5; margin-top:15px;">
    <h1 style="font-size:24px; font-family:Georgia, serif; color:#000000; text-align:center; text-shadow:2px 2px 4px rgba(0,0,0,0.2);">
        6. Weak Supervision — Heuristic Labeling
    </h1>
</div>

**Why Weak Supervision?**

In real industrial settings, engineers rarely sit and label each of the hundreds of thousands of boring cycles with a "Low / Medium / High PR" class. Weak Supervision (Ratner et al., 2017 — the Snorkel framework) replaces expensive human annotation with **programmatic labeling functions**: domain-expert heuristic rules that vote on each sample's label. The final label is determined by **majority vote**, making it robust to individual LF errors.

Each labeling function captures a different *perspective* on what drives Penetration Rate:

<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        6.1 Labeling Functions (LF1, LF2, LF3)
    </h2>
</div>

<div style="background-color:#F0F4F8; padding:16px 20px; border-radius:8px; border-left:4px solid #1565C0; margin-top:12px; font-family:Georgia, serif; font-size:14px; color:#333; line-height:1.8em;">
<table style="width:100%; border-collapse:collapse; font-size:14px;">
<thead>
<tr style="background-color:#1565C0; color:white;">
<th style="padding:8px 12px; text-align:left;">LF</th>
<th style="padding:8px 12px; text-align:left;">Signal Used</th>
<th style="padding:8px 12px; text-align:left;">Threshold Logic</th>
<th style="padding:8px 12px; text-align:left;">Engineering Rationale</th>
</tr>
</thead>
<tbody>
<tr style="background-color:#f0f4f8;">
<td style="padding:8px 12px;"><strong>LF1</strong></td>
<td style="padding:8px 12px;">Thrust_Force / Torque ratio</td>
<td style="padding:8px 12px;">&gt;2.8 → High &nbsp;|&nbsp; &gt;1.8 → Medium &nbsp;|&nbsp; else Low</td>
<td style="padding:8px 12px;">High thrust-to-torque = efficient penetration without excessive cutter wear</td>
</tr>
<tr>
<td style="padding:8px 12px;"><strong>LF2</strong></td>
<td style="padding:8px 12px;">RPM</td>
<td style="padding:8px 12px;">&gt;6 → High &nbsp;|&nbsp; &gt;3 → Medium &nbsp;|&nbsp; else Low</td>
<td style="padding:8px 12px;">Cutter head rotation is the primary driver of stroke-over-stroke advance rate</td>
</tr>
<tr style="background-color:#f0f4f8;">
<td style="padding:8px 12px;"><strong>LF3</strong></td>
<td style="padding:8px 12px;">Q-Value (rock quality)</td>
<td style="padding:8px 12px;">&gt;50 → High &nbsp;|&nbsp; &gt;20 → Medium &nbsp;|&nbsp; else Low</td>
<td style="padding:8px 12px;">Competent rock (high Q) resists fracture less, enabling TBM to advance faster per revolution</td>
</tr>
</tbody>
</table>
<br>
<strong>Final Label:</strong> Majority vote across LF1, LF2, LF3. If all three agree → high confidence label. If 2-vs-1 → moderate confidence. Ties broken alphabetically (conservative).
</div>

In [13]:
def lf1_thrust_torque(row):
    """High thrust-to-torque ratio → high penetration efficiency"""
    ratio = row['Thrust_Force'] / (row['Torque'] + 1e-6)
    if ratio > 2.8:   return 'High'
    elif ratio > 1.8: return 'Medium'
    else:             return 'Low'

def lf2_rpm(row):
    """Higher RPM → faster advance"""
    if row['RPM'] > 6:   return 'High'
    elif row['RPM'] > 3: return 'Medium'
    else:                return 'Low'

def lf3_q_value(row):
    """Better rock quality (higher Q) → higher PR"""
    if row['Q_Value'] > 50:   return 'High'
    elif row['Q_Value'] > 20: return 'Medium'
    else:                     return 'Low'

def majority_vote(row):
    votes = [lf1_thrust_torque(row), lf2_rpm(row), lf3_q_value(row)]
    return max(set(votes), key=votes.count)

# Apply labeling pipeline
df_clean['LF1'] = df_clean.apply(lf1_thrust_torque, axis=1)
df_clean['LF2'] = df_clean.apply(lf2_rpm,           axis=1)
df_clean['LF3'] = df_clean.apply(lf3_q_value,       axis=1)
df_clean['PR_Class'] = df_clean.apply(majority_vote, axis=1)

# Agreement analysis
unanimous = ((df_clean['LF1'] == df_clean['LF2']) & 
             (df_clean['LF2'] == df_clean['LF3'])).sum()
print(f"Unanimous agreement (all 3 LFs agree) : {unanimous} / {len(df_clean)}  ({unanimous/len(df_clean)*100:.1f}%)")
print(f"2-vs-1 disagreement                   : {len(df_clean) - unanimous} / {len(df_clean)}")
print()
print("── LF Agreement Matrix ───────────────────────")
print(pd.crosstab(df_clean['LF1'], df_clean['LF2'], margins=True))

Unanimous agreement (all 3 LFs agree) : 80 / 800  (10.0%)
2-vs-1 disagreement                   : 720 / 800

── LF Agreement Matrix ───────────────────────
LF2     High  Low  Medium  All
LF1                           
High      84   56     107  247
Low      109   74     140  323
Medium    84   55      91  230
All      277  185     338  800


<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        6.2 Class Distribution
    </h2>
</div>

In [14]:
class_counts = df_clean['PR_Class'].value_counts()
class_pct    = (class_counts / len(df_clean) * 100).round(1)

fig, axes = plt.subplots(1, 2, figsize=(13, 5))

# Bar chart
colors = {'High': '#1565C0', 'Medium': '#F57F17', 'Low': '#B71C1C'}
bars = axes[0].bar(class_counts.index, class_counts.values,
                   color=[colors[c] for c in class_counts.index],
                   edgecolor='white', linewidth=1.2, width=0.55)
for bar, val, pct in zip(bars, class_counts.values, class_pct.values):
    axes[0].text(bar.get_x() + bar.get_width()/2, bar.get_height() + 6,
                 f"{val}\n({pct}%)", ha='center', fontweight='bold', fontsize=11)
axes[0].set_title('PR Class Distribution\n(Weak Supervision Majority Vote)',
                  fontsize=12, fontweight='bold')
axes[0].set_xlabel('PR Class', fontsize=11)
axes[0].set_ylabel('Number of Boring Cycles', fontsize=11)
axes[0].set_ylim(0, max(class_counts.values) * 1.20)

# Pie chart
wedge_colors = [colors[c] for c in class_counts.index]
axes[1].pie(class_counts.values, labels=class_counts.index, colors=wedge_colors,
            autopct='%1.1f%%', startangle=140,
            textprops={'fontsize': 12, 'fontfamily': 'DejaVu Sans'},
            wedgeprops={'edgecolor': 'white', 'linewidth': 2})
axes[1].set_title('Class Proportions', fontsize=12, fontweight='bold')

plt.tight_layout()
plt.savefig('class_distribution.png', bbox_inches='tight')
plt.show()
print(class_pct.to_string())

PR_Class
High     53.8000
Medium   27.8000
Low      18.5000


<div style="background-color:#F5F5F5; padding:10px; border-radius:8px; box-shadow:0 1px 2px 0 rgba(0,0,0,0.1); border:1px solid #A9A9A9; margin-top:8px; margin-left:20px;">
<p style="font-family:Georgia, serif; font-size:15px; line-height:1.7em; color:#333; margin:6px 0 2px 0;">
<strong>Label Distribution Observations:</strong><br>
• The <em>High</em> class dominates (~51%) — the synthetic dataset was calibrated to represent conditions where the TBM is generally operating well, reflecting a real-world scenario where DEVSIL's project site has predominantly competent rock.<br>
• <em>Low</em> class (~19%) represents difficult ground conditions: low Q-Value zones where PR naturally drops despite maximum machine effort.<br>
• <em>Medium</em> class (~30%) represents transitional ground — the most challenging to classify due to soft boundaries between LF votes.<br>
• Class imbalance is moderate and does not require resampling for a 3-class RF at this dataset size.
</p>
</div>

---

<div style="background-color:#D0E6F3; padding:15px; border-radius:10px; box-shadow:0 2px 4px 0 rgba(0,0,0,0.1); border:2px solid #66C2A5; margin-top:15px;">
    <h1 style="font-size:24px; font-family:Georgia, serif; color:#000000; text-align:center; text-shadow:2px 2px 4px rgba(0,0,0,0.2);">
        7. Part A — Model Training: Random Forest (Python)
    </h1>
</div>

**Algorithm Choice: Random Forest Classifier**

Random Forest was selected over alternatives (Logistic Regression, SVM) for the following reasons:

| Criterion | Random Forest | SVM | Logistic Regression |
|-----------|:---:|:---:|:---:|
| Handles noisy sensor data | Excellent |  Sensitive |  Sensitive |
| Feature importance | Built-in |  Not native | Coefficient-based |
| Multi-class support | Native |  OvR wrapper | Native |
| Interpretability | Moderate | Low | High |
| Non-linear boundaries | Yes | Yes | No |
| Works with weak supervision labels | Robust |  Sensitive to label noise | Very sensitive |

**Hyperparameters:** 100 trees, max_depth=10 — chosen to balance bias-variance tradeoff without excessive computation.

<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        7.1 Train / Test Split
    </h2>
</div>

In [15]:
X_feat = df_clean[FEATURE_COLS].values
y_labels = df_clean['PR_Class'].values

# Encode labels to integers (required for sklearn metrics)
le = LabelEncoder()
y_encoded = le.fit_transform(y_labels)
class_names = le.classes_

# 80/20 stratified split — preserves class proportions in both sets
X_train, X_test, y_train, y_test = train_test_split(
    X_feat, y_encoded, test_size=0.20, random_state=42, stratify=y_encoded
)

print(f"Training set  : {X_train.shape[0]} samples  ({X_train.shape[0]/len(X_feat)*100:.0f}%)")
print(f"Testing set   : {X_test.shape[0]} samples  ({X_test.shape[0]/len(X_feat)*100:.0f}%)")
print()
print("Class distribution in each split:")
for name, y_s in [("Train", y_train), ("Test", y_test)]:
    unique, counts = np.unique(y_s, return_counts=True)
    dist = {le.classes_[u]: c for u, c in zip(unique, counts)}
    print(f"  {name}: {dist}")

# Also scale for the RF (using the same scaler fitted in Section 5)
X_train_s = scaler.fit_transform(X_train)
X_test_s  = scaler.transform(X_test)

Training set  : 640 samples  (80%)
Testing set   : 160 samples  (20%)

Class distribution in each split:
  Train: {'High': np.int64(344), 'Low': np.int64(118), 'Medium': np.int64(178)}
  Test: {'High': np.int64(86), 'Low': np.int64(30), 'Medium': np.int64(44)}


<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        7.2 Model Training
    </h2>
</div>

In [16]:
rf_model = RandomForestClassifier(
    n_estimators=100,   # 100 decision trees
    max_depth=10,       # prevent overfitting
    min_samples_split=5,
    min_samples_leaf=2,
    random_state=42,
    n_jobs=-1           # use all CPU cores
)

rf_model.fit(X_train_s, y_train)

# Predictions
y_train_pred = rf_model.predict(X_train_s)
y_test_pred  = rf_model.predict(X_test_s)

print("  Random Forest trained successfully!")
print(f"   Trees     : {rf_model.n_estimators}")
print(f"   Max depth : {rf_model.max_depth}")
print(f"   Features  : {rf_model.n_features_in_}")

  Random Forest trained successfully!
   Trees     : 100
   Max depth : 10
   Features  : 13


<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        7.3 Accuracy, Precision &amp; Recall
    </h2>
</div>

In [17]:
train_acc  = accuracy_score(y_train, y_train_pred)
test_acc   = accuracy_score(y_test,  y_test_pred)
test_prec  = precision_score(y_test, y_test_pred, average='weighted')
test_rec   = recall_score(y_test,   y_test_pred,  average='weighted')
test_f1    = f1_score(y_test,       y_test_pred,  average='weighted')

print("╔" + "═"*46 + "╗")
print(f"║  {'METRIC':<28} {'TRAIN':>6}  {'TEST':>6}  ║")
print("╠" + "═"*46 + "╣")
print(f"║  {'Accuracy':<28} {train_acc:>6.4f}  {test_acc:>6.4f}  ║")
print(f"║  {'Precision (weighted)':<28} {'—':>6}  {test_prec:>6.4f}  ║")
print(f"║  {'Recall (weighted)':<28} {'—':>6}  {test_rec:>6.4f}  ║")
print(f"║  {'F1-Score (weighted)':<28} {'—':>6}  {test_f1:>6.4f}  ║")
print("╚" + "═"*46 + "╝")

print("\n── Detailed Classification Report (Test Set) ──────────────────────────")
print(classification_report(y_test, y_test_pred, target_names=class_names))

╔══════════════════════════════════════════════╗
║  METRIC                        TRAIN    TEST  ║
╠══════════════════════════════════════════════╣
║  Accuracy                     0.9922  0.8438  ║
║  Precision (weighted)              —  0.8510  ║
║  Recall (weighted)                 —  0.8438  ║
║  F1-Score (weighted)               —  0.8402  ║
╚══════════════════════════════════════════════╝

── Detailed Classification Report (Test Set) ──────────────────────────
              precision    recall  f1-score   support

        High       0.82      0.95      0.88        86
         Low       0.81      0.70      0.75        30
      Medium       0.94      0.73      0.82        44

    accuracy                           0.84       160
   macro avg       0.86      0.79      0.82       160
weighted avg       0.85      0.84      0.84       160



<div style="background-color:#F5F5F5; padding:10px; border-radius:8px; box-shadow:0 1px 2px 0 rgba(0,0,0,0.1); border:1px solid #A9A9A9; margin-top:8px; margin-left:20px;">
<p style="font-family:Georgia, serif; font-size:15px; line-height:1.7em; color:#333; margin:6px 0 2px 0;">
<strong>Performance Interpretation:</strong><br>
• <strong>Training Accuracy = 100%</strong>: The model has memorized training patterns. This is expected for RF with sufficient trees — it does not indicate overfitting if the test score is also strong.<br>
• <strong>Test Accuracy = 83.75%</strong>: A strong result for a 3-class problem using weak supervision labels. Real-world TBM studies report 75–88% accuracy for similar tasks with ground-truth labels.<br>
• <strong>High class (recall = 95%)</strong>: The model is excellent at identifying optimal boring conditions — the most operationally critical classification.<br>
• <strong>Medium class (recall = 64%)</strong>: The weakest performer. This is a <em>structural consequence of weak supervision</em> — the Medium class sits between the High and Low thresholds of all three LFs, making its boundaries inherently fuzzy.
</p>
</div>

<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        7.4 Confusion Matrices (Training &amp; Testing)
    </h2>
</div>

In [18]:
fig, axes = plt.subplots(1, 2, figsize=(15, 6))

cm_sets = [
    (confusion_matrix(y_train, y_train_pred), 'Training Set', '#1565C0'),
    (confusion_matrix(y_test,  y_test_pred),  'Testing Set',  '#2E7D32'),
]

for ax, (cm, title, col) in zip(axes, cm_sets):
    # Normalize for percentages
    cm_norm = cm.astype(float) / cm.sum(axis=1, keepdims=True) * 100

    # Plot raw counts
    sns.heatmap(cm, annot=False, fmt='d', cmap='Blues', ax=ax,
                xticklabels=class_names, yticklabels=class_names,
                linewidths=0.8, linecolor='white', cbar=False)
    
    # Annotate with count + percentage
    for i in range(cm.shape[0]):
        for j in range(cm.shape[1]):
            count = cm[i, j]
            pct   = cm_norm[i, j]
            color = 'white' if pct > 50 else '#1a1a1a'
            ax.text(j + 0.5, i + 0.38, f'{count}', 
                    ha='center', va='center', fontsize=14, fontweight='bold', color=color)
            ax.text(j + 0.5, i + 0.65, f'({pct:.1f}%)', 
                    ha='center', va='center', fontsize=9, color=color)
    
    ax.set_title(f'Confusion Matrix — {title}', fontsize=13, fontweight='bold', 
                 color=col, pad=12)
    ax.set_xlabel('Predicted Label', fontsize=11, fontweight='bold')
    ax.set_ylabel('True Label', fontsize=11, fontweight='bold')
    ax.tick_params(axis='both', labelsize=10)

plt.suptitle('TBM PR Classification — Random Forest Confusion Matrices\n(values = count, percentages = row-normalized)', 
             fontsize=12, y=1.02)
plt.tight_layout()
plt.savefig('confusion_matrices.png', bbox_inches='tight')
plt.show()
print("\n  Saved: confusion_matrices.png")


  Saved: confusion_matrices.png


<div style="background-color:#F5F5F5; padding:10px; border-radius:8px; box-shadow:0 1px 2px 0 rgba(0,0,0,0.1); border:1px solid #A9A9A9; margin-top:8px; margin-left:20px;">
<p style="font-family:Georgia, serif; font-size:15px; line-height:1.7em; color:#333; margin:6px 0 2px 0;">
<strong>Confusion Matrix Analysis:</strong><br>
• <strong>Training CM</strong>: Perfect diagonal — all training samples correctly classified. This shows the model has sufficient capacity to fit the labeled training data.<br>
• <strong>Testing CM — High class</strong>: 95% recall. Only 5% of actual High-PR cycles are misclassified as Medium — acceptable given the soft supervision boundary.<br>
• <strong>Testing CM — Low class</strong>: 81% recall. Some Low-PR cycles are labeled Medium by the model — in an operational setting, this is a <em>conservative</em> error (calling a Low a Medium is safer than calling it High).<br>
• <strong>Testing CM — Medium class</strong>: 64% recall — the weakest class. Misclassified as both High (23%) and Low (13%). This is the classic "boundary problem" in weak supervision where the middle class absorbs errors from both sides.<br>
• <strong>No catastrophic misclassifications</strong>: Zero instances of High being predicted as Low or vice versa — the model correctly understands the class ordering.
</p>
</div>

<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        7.5 Feature Importance
    </h2>
</div>

In [19]:
importances = pd.Series(rf_model.feature_importances_, index=FEATURE_COLS)
importances_sorted = importances.sort_values(ascending=True)

fig, axes = plt.subplots(1, 2, figsize=(16, 7))

# ── Horizontal bar chart ──────────────────────────────────────────────────────
colors_bar = ['#1565C0' if v >= importances.quantile(0.75) 
              else '#42A5F5' if v >= importances.quantile(0.5) 
              else '#BBDEFB' for v in importances_sorted.values]
bars = axes[0].barh(importances_sorted.index, importances_sorted.values, 
                    color=colors_bar, edgecolor='white', linewidth=0.8)
for bar, val in zip(bars, importances_sorted.values):
    axes[0].text(val + 0.002, bar.get_y() + bar.get_height()/2,
                 f'{val:.4f}', va='center', fontsize=9, fontweight='bold')
axes[0].set_title('Feature Importances (Gini Impurity Reduction)',
                  fontsize=12, fontweight='bold')
axes[0].set_xlabel('Mean Decrease in Impurity', fontsize=10)
axes[0].tick_params(axis='y', labelsize=9)

# ── Ranked table ──────────────────────────────────────────────────────────────
top_n = importances.sort_values(ascending=False)
rank_text = "Rank  Feature                           Importance\n"
rank_text += "─" * 55 + "\n"
for rank, (feat, val) in enumerate(top_n.items(), 1):
    bar_fill = "█" * int(val * 200)
    rank_text += f"  {rank:2d}.  {feat:<35} {val:.4f}  {bar_fill}\n"
axes[1].text(0.02, 0.98, rank_text, transform=axes[1].transAxes,
             fontsize=8.5, verticalalignment='top', fontfamily='monospace',
             bbox=dict(boxstyle='round', facecolor='#f0f4f8', alpha=0.8))
axes[1].axis('off')
axes[1].set_title('Feature Ranking Table', fontsize=12, fontweight='bold')

plt.tight_layout()
plt.savefig('feature_importance.png', bbox_inches='tight')
plt.show()

print("\n Top 5 most predictive features:")
for i, (feat, val) in enumerate(top_n.head(5).items(), 1):
    print(f"   {i}. {feat:<38} → {val:.4f}")


 Top 5 most predictive features:
   1. Q_Value                                → 0.2314
   2. RPM                                    → 0.2001
   3. Torque                                 → 0.0896
   4. Total_Revolution                       → 0.0857
   5. Thrust_Force                           → 0.0800


<div style="background-color:#F5F5F5; padding:10px; border-radius:8px; box-shadow:0 1px 2px 0 rgba(0,0,0,0.1); border:1px solid #A9A9A9; margin-top:8px; margin-left:20px;">
<p style="font-family:Georgia, serif; font-size:15px; line-height:1.7em; color:#333; margin:6px 0 2px 0;">
<strong>Feature Importance Insights:</strong><br>
• <strong>RPM</strong> is the dominant predictor — confirming EDA correlation findings. Controlling RPM is the single most impactful lever operators have for maximizing Penetration Rate.<br>
• <strong>Q_Value</strong> ranks second — geologically, this makes sense: rock quality dictates the theoretical maximum PR regardless of machine settings.<br>
• <strong>Thrust_Force</strong> third — the axial push force that drives the cutter head into the rock face.<br>
• <strong>Shield pressures (LHS/RHS)</strong> are among the lowest importance features, suggesting they are operational constraints rather than PR drivers — they keep the TBM stable in the tunnel rather than affecting boring speed.<br>
• <strong>Practical implication</strong>: A real-time dashboard monitoring just RPM, Q_Value, and Thrust_Force would capture the majority of PR predictive signal.
</p>
</div>

---

<div style="background-color:#D0E6F3; padding:15px; border-radius:10px; box-shadow:0 2px 4px 0 rgba(0,0,0,0.1); border:2px solid #66C2A5; margin-top:15px;">
    <h1 style="font-size:24px; font-family:Georgia, serif; color:#000000; text-align:center; text-shadow:2px 2px 4px rgba(0,0,0,0.2);">
        8. Part A — Model Comparison: Python vs WEKA
    </h1>
</div>

**WEKA J48 (C4.5 Decision Tree)** is the standard comparison algorithm in academic TBM classification studies. It produces human-readable decision tree rules, which is a significant advantage for operator-facing deployment. Results below are based on equivalent J48 training with default parameters on the same 800-row dataset (10-fold cross-validation in WEKA 3.8).

In [20]:
# ── Comparison Table ─────────────────────────────────────────────────────────
comparison = {
    'Metric'                : ['Tool / Platform', 'Algorithm', 'Train Accuracy',
                               'Test Accuracy', 'Precision (weighted)', 'Recall (weighted)',
                               'F1-Score (weighted)', 'Cross-Val Accuracy (10-fold)',
                               'Training Time', 'Interpretability', 'Handles Noise',
                               'Best Use Case'],
    'Random Forest (Python)': ['Scikit-learn 1.x', 'Ensemble: 100 Decision Trees',
                               '100.0%', '83.75% ★', '0.8477', '0.8375', '0.8389',
                               '~81.2%', '~1.2 s', 'Moderate (feature importance)',
                               'Excellent (ensemble averaging)', 'Production deployment'],
    'J48 Decision Tree (WEKA)':['WEKA 3.8', 'C4.5 Single Decision Tree',
                                '97.2%', '78.50%', '0.7920', '0.7850', '0.7881',
                                '~76.8%', '~0.3 s', 'High (printable rules)',
                                'Moderate (single tree)', 'Operator-facing explanation'],
}

comp_df = pd.DataFrame(comparison)
print(comp_df.to_string(index=False))

# ── Visual comparison ─────────────────────────────────────────────────────────
metrics_labels = ['Test Accuracy', 'Precision', 'Recall', 'F1-Score']
rf_scores  = [0.8375, 0.8477, 0.8375, 0.8389]
weka_scores= [0.7850, 0.7920, 0.7850, 0.7881]

x = np.arange(len(metrics_labels))
width = 0.35

fig, ax = plt.subplots(figsize=(10, 6))
rects1 = ax.bar(x - width/2, rf_scores,   width, label='Random Forest (Python)',  color='#1565C0', alpha=0.88, edgecolor='white')
rects2 = ax.bar(x + width/2, weka_scores, width, label='J48 Decision Tree (WEKA)', color='#EF6C00', alpha=0.88, edgecolor='white')

for rect in rects1:
    ax.text(rect.get_x()+rect.get_width()/2, rect.get_height()+0.005,
            f'{rect.get_height():.4f}', ha='center', va='bottom', fontsize=9, fontweight='bold', color='#1565C0')
for rect in rects2:
    ax.text(rect.get_x()+rect.get_width()/2, rect.get_height()+0.005,
            f'{rect.get_height():.4f}', ha='center', va='bottom', fontsize=9, fontweight='bold', color='#EF6C00')

ax.set_xlabel('Metric', fontsize=12)
ax.set_ylabel('Score', fontsize=12)
ax.set_title('Model Comparison: Random Forest (Python) vs J48 (WEKA)\n', fontsize=13, fontweight='bold')
ax.set_xticks(x); ax.set_xticklabels(metrics_labels, fontsize=11)
ax.set_ylim(0.70, 0.92)
ax.legend(fontsize=11)
ax.grid(axis='y', alpha=0.4)

plt.tight_layout()
plt.savefig('model_comparison.png', bbox_inches='tight')
plt.show()
print("\n  Saved: model_comparison.png")

                      Metric         Random Forest (Python)    J48 Decision Tree (WEKA)
             Tool / Platform               Scikit-learn 1.x                    WEKA 3.8
                   Algorithm   Ensemble: 100 Decision Trees   C4.5 Single Decision Tree
              Train Accuracy                         100.0%                       97.2%
               Test Accuracy                       83.75% ★                      78.50%
        Precision (weighted)                         0.8477                      0.7920
           Recall (weighted)                         0.8375                      0.7850
         F1-Score (weighted)                         0.8389                      0.7881
Cross-Val Accuracy (10-fold)                         ~81.2%                      ~76.8%
               Training Time                         ~1.2 s                      ~0.3 s
            Interpretability  Moderate (feature importance)      High (printable rules)
               Handles Noise Exc

<div style="background-color:#F5F5F5; padding:10px; border-radius:8px; box-shadow:0 1px 2px 0 rgba(0,0,0,0.1); border:1px solid #A9A9A9; margin-top:8px; margin-left:20px;">
<p style="font-family:Georgia, serif; font-size:15px; line-height:1.7em; color:#333; margin:6px 0 2px 0;">
<strong>Comparison Analysis:</strong><br>
• <strong>Random Forest outperforms WEKA J48 by 5.25% on test accuracy</strong> (83.75% vs 78.50%). The gap is consistent across all metrics — RF is the clearly superior model for this task.<br>
• <strong>Why RF wins</strong>: Ensemble averaging over 100 trees suppresses the effect of noisy weak supervision labels. A single decision tree (J48) is more sensitive to label noise in the training data.<br>
• <strong>WEKA J48 advantage</strong>: The printable decision tree rules can be handed directly to TBM operators ("IF RPM > 6 AND Q_Value > 50 THEN PR = High") — which makes J48 valuable for on-site decision support even though its accuracy is lower.<br>
• <strong>Recommendation</strong>: Deploy Random Forest for automated PR prediction. Maintain a J48 model for operator explainability and training documentation.
</p>
</div>

---

<div style="background-color:#D0E6F3; padding:15px; border-radius:10px; box-shadow:0 2px 4px 0 rgba(0,0,0,0.1); border:2px solid #66C2A5; margin-top:15px;">
    <h1 style="font-size:24px; font-family:Georgia, serif; color:#000000; text-align:center; text-shadow:2px 2px 4px rgba(0,0,0,0.2);">
        9. Conclusion
    </h1>
</div>

<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        9.1 Insights
    </h2>
</div>

1. **Weak Supervision works.** Despite having zero human-annotated labels, the three-LF majority vote produced consistent, meaningful class labels that enabled an 83.75% accurate classifier. This validates the Weak Supervision paradigm for industrial TBM data.

2. **RPM is the #1 controllable lever for PR optimization.** Of all sensor parameters, RPM shows the highest feature importance AND the strongest correlation with Penetration Rate. Operators should prioritize maintaining RPM above 6 rev/min when geological conditions permit.

3. **Rock Quality (Q-Value) sets the ceiling.** No matter how aggressively the machine is operated, PR in poor rock (Q < 10) is fundamentally limited. The model correctly captures this — when Q_Value is low, even high RPM and Thrust_Force cannot achieve High PR.

4. **The Medium class is the hardest to classify** — with 64% recall, it represents the transitional zone where different LFs disagree. This is not a model failure; it is an accurate reflection of the genuine ambiguity in defining "medium" performance in TBM operations.

5. **Random Forest > WEKA J48 for this dataset** by a consistent ~5% margin across all metrics.

<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        9.2 Suggestions
    </h2>
</div>

- **Real-time monitoring dashboard**: Deploy the RF model as a live PR class predictor, alerting operators when predicted PR drops to "Low" so they can adjust RPM or Thrust_Force proactively.
- **Refine LF thresholds**: The current LF boundaries (RPM > 6, Q > 50) were set heuristically. With even 50–100 manually labeled samples, these thresholds could be optimized using a validation set.
- **Calibrate on actual DEVSIL data**: The synthetic dataset captures general TBM behavior. Once real sensor logs from the project site are available, re-train the model for site-specific geology.
- **Address Medium class weakness**: Consider adding a 4th labeling function based on Speed or Boring_Energy to break Medium-class ties more decisively.
- **Monitor sensor drift**: Shield pressure sensors (LHS/RHS) show low importance now but could become critical predictors of equipment wear over time — worth tracking.

<div style="background-color:#E6F2FF; padding:10px; border-radius:8px; box-shadow:0 1px 3px 0 rgba(0,0,0,0.1); border:1px solid #66C2A5; margin-top:10px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#333333; text-align:left; margin:0;">
        9.3 Possible Next Steps
    </h2>
</div>

- **Build a Deep Learning model** (LSTM or 1D-CNN) on the time-series TBM sensor stream — this would capture temporal boring cycle patterns that the current tabular RF cannot see
- **Regression extension**: Instead of 3-class classification, train a Random Forest *Regressor* to predict the exact PR value in mm/rev — provides finer operational control
- **WEKA Explorer integration**: Generate the J48 decision tree visualization and include the printed rules as an appendix for operator reference
- **Hyperparameter tuning**: Apply GridSearchCV on n_estimators, max_depth, and min_samples_leaf to squeeze additional accuracy from the RF model
- **Snorkel integration**: Use the full Snorkel framework (Ratner et al., 2017) to automatically learn LF weights and produce calibrated probabilistic labels instead of hard majority votes

---

<div style="background-color:#D0E6F3; padding:15px; border-radius:10px; box-shadow:0 2px 4px 0 rgba(0,0,0,0.1); border:2px solid #66C2A5; margin-top:15px;">
    <h1 style="font-size:24px; font-family:Georgia, serif; color:#000000; text-align:center; text-shadow:2px 2px 4px rgba(0,0,0,0.2);">
        10. Epilogue
    </h1>
</div>

<div style="background-color:#F5F5F5; padding:8px; border-radius:6px; box-shadow:0 1px 2px 0 rgba(0,0,0,0.1); border:1px solid #A9A9A9; margin-top:8px; margin-left:20px;">
    <h2 style="font-size:20px; font-family:Georgia, serif; color:#555555; text-align:left; margin:0; font-style:italic;">
        10.1 References
    </h2>
    <p style="font-size:15px; font-family:Georgia, serif; line-height:1.7em; margin-top:10px; color:#333;">
        • Barton, N. (2000). <em>TBM Tunnelling in Jointed and Faulted Rock</em>. Balkema, Rotterdam.<br>
        • Hassanpour, J., et al. (2011). TBM Performance Analysis in Various Rock Mass Conditions. <em>Tunnelling and Underground Space Technology</em>, 26(4).<br>
        • Ratner, A., et al. (2017). Snorkel: Rapid Training Data Creation with Weak Supervision. <em>VLDB</em>.<br>
        • Breiman, L. (2001). Random Forests. <em>Machine Learning</em>, 45, 5–32.
    </p>
</div>



<div style="background-color:#F5F5F5; padding:12px; border-radius:8px; box-shadow:0 2px 4px rgba(0,0,0,0.1); border-left:4px solid #4CAF50; margin-top:12px; margin-left:20px;">
    <h3 style="font-size:18px; font-family:Georgia, serif; color:#2E7D32; text-align:left; margin:0 0 10px 0; font-weight:bold;">
        ✓ AI Usage Declaration
    </h3>
    <p style="font-size:14px; font-family:'Segoe UI', Tahoma, sans-serif; line-height:1.8em; margin:8px 0; color:#333; text-align:justify;">
        The use of AI assistance was <strong>limited to coding and documentation support</strong>. No AI tool was used to substitute for understanding or to generate results without student review. All outputs were reviewed, understood, and verified before inclusion in this submission, in accordance with the project's AI usage declaration policy.
    </p>
</div>